pkgs/build-support/docker/default.nix

```diff
@@ -30,7 +30,20 @@ rec {
     inherit pkgs buildImage pullImage shadowSetup;
   };
 
-  pullImage = callPackage ./pull.nix {};
+  pullImage =
+    let
+      nameReplace = name: builtins.replaceStrings ["/" ":"] ["-" "-"] name;
+    in
+      # For simplicity we only support sha256.
+      { imageName, imageTag ? "latest", imageId ? "${imageName}:${imageTag}"
```

Comment:
Does "latest" really make sense as a tag to use here? I know you're just carrying it over from the previous version, but it seems like a recipe for distributing unreproducible code: I use `imageTag = "latest";`, set the sha256 I get, and then as long as it stays in my store, I don't think about it for a while. I push up the code to my colleagues who don't share my store, and they pull `latest` and get a different image. If anything I'd try to actively discourage folks from using `latest`. That might be a topic for a separate review though 😄

Reply:
I fully agree with you.
Since this PR will be backported into 17.09, maybe it would be better to minimize changes.

So I could open another PR to remove the default "latest" value and go further. 
It's possible to pull image by using the image digest (immutable):
https://docs.docker.com/engine/reference/commandline/pull/#pull-an-image-by-digest-immutable-identifier
Moreover, this is supported by Skopeo. So maybe we could add a `digest` argument and print a warning if the user uses the tag. I'm not sure that removing the tag argument is a good idea because it's not easy to find the digest of an image (I was not able to find it on the docker hub...).

Reply:
Being able to pull by the image digest, previously called `imageId` in `dockerTools`, is what my team has been using previously. The change earlier this year repurposed `imageId` to what it is now, which is not nearly as useful and has prevented us from updating `nixpkgs` on our projects. I am in favor of changing `imageId` back to what it meant before the change use a VM.

Reply:
@purefn is there an open issue for that? We shouldn't be breaking interfaces without providing reasonable alternatives, at least

Reply:
@copumpkin Sadly, no. I had run into another issue with that commit, #27294, and thought when that had been fixed it fixed our only issue. I tried it and encountered more problems, but didn't  get to in any further until just recently.

Reply:
Can we come to a decision and get this merged and backported? My $0.02 is I think it is important to be able to pull by digest. This tooling has been largely broken forever. I'm not convinced too many people are depending on it.

Reply:
@grahamc we depend on digests but I think we agree that this PR is better than the current state.

Reply:
I think this is not relevant to this improvement.

Reply:
Thanks.
And I've open #29505 to continue the discussion :)